### ISSUE TITLE
`@xstate/inspect` not updating in CodeSandbox when `react@>=16.9.0`, `react-dom@>=16.9.0`

### ISSUE THREAD

**Issue by with-heart** [davidkpiano/xstate#1696]
**Description**
This is an issue I've been noticing for awhile that I finally tracked down to what appears to be the `react` version being used. For the longest time when using `@xstate/inspect` with React, `@xstate/inspect` wouldn't update when I'd save changes to the machine. I had to manually refresh the entire CodeSandbox and allow it to open a new inspector window to see any updates.

**Expected Result**
When saving changes to a machine being used with `useMachine(machine, {devTools: true})`, the inspector updates to show the changes.

**Actual Result**
The inspector doesn't update when I save changes to the machine.

**Reproduction**
Working CodeSandbox: https://codesandbox.io/s/lucid-ellis-vsn3x?file=/src/App.js

This example is using `react@16.8.6` and `react-dom@16.8.6`. When you save changes to `toggleMachine`, the inspector correctly updates the machine. You can see the inspector page refresh entirely and afterwards the updated machine is shown.

Broken CodeSandbox: https://codesandbox.io/s/serene-butterfly-y120n?file=/src/App.js

This example is using `react@16.9.0` and `react-dom@16.9.0`. It's identical to the working example in all ways except for the versions. When you save changes to the machine in this example, the inspector does not update. You can see the machine title (`toggle (x:1)`) flash very briefly when you save, but the inspector page doesn't refresh and no changes are shown.

**Additional context**
I tried to run both of these locally to see if the issue is particular to CodeSandbox and interestingly enough the example that works in CodeSandbox doesn't work at all locally. The machine title doesn't even flash. I'm really not sure what to make of that.

**Comment by davidkpiano** [comment]
This might be a problem with our use of `useConstant(...)` to create the machine (cc. @Andarist) - it holds a (potentially lazy) reference to a machine that lives forever, even through fast refresh, apparently.

This also might leak into the other problem re: changing the `machine` passed into `useMachine(machine)`, and what the expected behavior is there. Currently, we _ignore_ any new machines.

**Comment by Andarist** [comment]
Yep - this is an issue with Fast Refresh itself. It's currently broken in multiple ways with how `useMachine` works - adding `@xstate/inspect` into the mix only makes things worse.

The ideal fix for us would be if we could detect that Fast Refresh is happening so we could handle this in a special manner during dev environments. I'll be trying to chat with React team about exposing this information on the Fast Refresh runtime.

**Comment by ericclemmons** [comment]
> This also might leak into the other problem re: changing the `machine` passed into `useMachine(machine)`, and what the expected behavior is there. Currently, we _ignore_ any new machines.

@Andarist Is the ideal to not ignore new machines, but replace/update them? In product, this wouldn't typically happen (unless there's a dynamic state generator?), but updating when machine definitions/references change would be least surprising, right?

**Comment by Andarist** [comment]
From our PoV the machine definition is/should be static and we treat it as such. It's like a formalized behavior of your component and that shouldn't generally be dynamic. There are uneasy questions that would have to be answered to allow for this (for example - what happens with currently invoked services?) and we don't really think that those are valid questions to ask in the first place.

It's also not like this is creating precedence - those are already there 😅 Just think of `useMachine` as of `useState`'s initializer function. Changing it won't affect your state anyhow, right? 😉

**Comment by Andarist** [comment]
I believe that the situation around this got substantially improved by https://github.com/statelyai/xstate/pull/4436 for the upcoming v5 release

The inspector might still need some follow-up work to handle this properly but I think that the "core" issue has been solved and that we can track further improvements separately.
